File: CandidateSolution/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> result;
    vector<int> temp;
    
    void recur(vector<int>& candidates, int target, int i) {
            if (target < 0) {
                return;
            }
            
            if (target == 0) {
                result.push_back(temp);
                return;
            }
            
            while (i < candidates.size() && target - candidates[i] >= 0) {
                temp.push_back(candidates[i]);
                
                recur(candidates, target - candidates[i], i);
                ++i;
                temp.pop_back();
                
            }
        
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        recur(candidates, target, 0);
        return result;
    
    }
};

File: java/39. Combination Sum.java
// 39. Combination Sum
// TC: O(N^(T/M)) where N = candidates length, T = target, M = minimal candidate
// SC: O(T/M) for the recursion stack and the space to store results
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> lt = new ArrayList<List<Integer>>();
        List<Integer> subarray = new ArrayList<Integer>();
        int idx = 0;
        backtracking(idx, candidates, target, subarray, lt);
        return lt;
    }

    private void backtracking(int idx, int[] candidates, int target, List<Integer> sub,
        List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<>(sub));
            return;
        } else if (target < 0) {
            return;
        }

        for (int i = idx; i < candidates.length && candidates[i] <= target; i++) {
            sub.add(candidates[i]);
            backtracking(i, candidates, target - candidates[i], sub, res);
            sub.remove(sub.size() - 1);
        }
    }
}